### PR TITLE
Remove dead Boehm code.

### PR DESCRIPTION
**Pull request by jaykrell**
Unity looked similarly dead.

The key points making it dead were:

create_allocator:
  g_assert_not_reached() covering everything but local variables

mono_gc_get_managed_allocator:
  always returning NULL, killing a bunch of code
  after it

mono_gc_is_critical_method:
   looking at data that was never filled in

This should have no impact, not even on Boehm.
Boehm should continue working just the same.

It is nice because it is the few/only uses of CEE_MONO_TLS.

### CODE REVIEW

File: mono/metadata/boehm-gc.c [hunk 1]
```diff
@@ -959,371 +959,9 @@ mono_gc_get_restart_signal (void)
...
-	int index_var, bytes_var, my_fl_var, my_entry_var;
-	guint32 no_freelist_branch, not_small_enough_branch = 0;
-	guint32 size_overflow_branch = 0;
-	MonoMethodBuilder *mb;
-	MonoMethod *res;
-	MonoMethodSignature *csig;
-	const char *name = NULL;
-	WrapperInfo *info;
-
-	g_assert_not_reached ();
```
Comment (jaykrell) [inline mono/metadata/boehm-gc.c]: Here.

File: mono/metadata/boehm-gc.c [hunk 2]
```diff
@@ -959,371 +959,9 @@ mono_gc_get_restart_signal (void)
...
-MonoMethod*
-mono_gc_get_managed_allocator (MonoClass *klass, gboolean for_box, gboolean known_instance_size)
-{
-	int atype;
-
-	/*
-	 * Tls implementation changed, we jump to tls native getters/setters.
-	 * Is boehm managed allocator ok with this ? Do we even care ?
-	 */
-	return NULL;
```
Comment (jaykrell) [inline mono/metadata/boehm-gc.c]: Here.

File: mono/metadata/boehm-gc.c [hunk 3]
```diff
@@ -959,371 +959,9 @@ mono_gc_get_restart_signal (void)
...
- *   Return a managed allocator method corresponding to allocator type ATYPE.
- */
-MonoMethod*
-mono_gc_get_managed_allocator_by_type (int atype, ManagedAllocatorVariant variant)
-{
-	MonoMethod *res;
-	gboolean slowpath = variant != MANAGED_ALLOCATOR_REGULAR;
-	MonoMethod **cache = slowpath ? slowpath_alloc_method_cache : alloc_method_cache;
-
-	return NULL;
```
Comment (jaykrell) [inline mono/metadata/boehm-gc.c]: Here.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by BrzVlad** [comment]
@joncham Are you aware that this code is dead ? I am assuming it has a preaty heavy perf impact and it wouldn't be too hard to fix (assuming it is only related to tls access)

**Comment by jaykrell** [comment]
Leave it for restoration?

**Comment by jaykrell** [comment]
How about instead I `#if 0` it out?

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by lambdageek** [comment]
Assuming @joncham doesn't want this code, I'd prefer to delete rather than `#if 0`

**Comment by jaykrell** [comment]
Agreed, and let me trim at the ‘#if’, will clean up later.

File: mono/metadata/boehm-gc.c [hunk 4]
```diff
@@ -959,371 +959,9 @@ mono_gc_get_restart_signal (void)
 }
 
 #if defined(USE_COMPILER_TLS) && defined(__linux__) && (defined(__i386__) || defined(__x86_64__))
```
Comment (jaykrell) [inline mono/metadata/boehm-gc.c]: Remove from ‘#if’ to/including ‘#else’ instead.

Comment (jaykrell) [reply mono/metadata/boehm-gc.c]: Now that I have bigger screen -- this isn't terrible. Just a tiny excess remains.

**Comment by jaykrell** [comment]
Pinged Jon here and in slack, no answer.

**Pull request merged**